File: Android/indoor_localization/lib/services/asset_service.dart
import 'package:indoor_localization/entities/asset.dart';
import 'package:indoor_localization/entities/facility.dart';
import 'package:indoor_localization/entities/point.dart';
import 'package:indoor_localization/entities/zone.dart';
import 'package:indoor_localization/services/entity_service.dart';

class AssetService extends EntityService {
  static Asset? get(int id) {
    return EntityService.assetRepo.get(id);
  }

  static List<Asset> getAll() {
    return EntityService.assetRepo.getAll();
  }

  static bool add(Asset asset) {
    return EntityService.assetRepo.add(asset);
  }

  static bool delete(int id) {
    return EntityService.assetRepo.delete(id);
  }

  static bool update(Asset asset) {
    return EntityService.assetRepo.update(asset);
  }

  static Facility? getAssetParentFacility(int id) {
    final asset = EntityService.assetRepo.get(id);
    if (asset != null) {
      return EntityService.facilityRepo.get(asset.parentFacilityId);
    }
    return null;
  }

  static Point? getAssetPosition(int id) {
    final asset = EntityService.assetRepo.get(id);
    if (asset != null) {
      return asset.position;
    }
    return null;
  }

  static DateTime? getAssetLastSync(int id) {
    final asset = EntityService.assetRepo.get(id);
    if (asset != null) {
      return asset.lastSync;
    }
    return null;
  }

  static bool isAssetActive(int id) {
    final asset = EntityService.assetRepo.get(id);
    return asset != null ? asset.isActive : false;
  }

  static List<Zone> getAssetCurrentZones(int id) {
    final asset = EntityService.assetRepo.get(id);
    if (asset != null) {
      return asset.currentZonesIds
          .map((zoneId) => EntityService.zoneRepo.get(zoneId))
          .where((zone) => zone != null)
          .cast<Zone>()
          .toList();
    }
    return [];
  }
}

File: WebApp/indoor-localization/src/services/AssetService.ts
import { Asset } from "../entities/Asset";
import { Facility } from "../entities/Facility";
import { Zone } from "../entities/Zone";
import { FacilityService } from "./FacilityService";
import { EntityService } from "./Service";
import { ZoneService } from "./ZoneService";

export class AssetService extends EntityService {
    static async Get(id: number): Promise<Asset | null> {
        return this.assetRepo.Get(id);
      }

    static async GetAll(): Promise<Asset[]> {
        return this.assetRepo.GetAll();
    }

    static async Add(asset: Asset): Promise<boolean> {
        return this.assetRepo.Add(asset);
    }

    static async Delete(id: number): Promise<boolean> {
        return this.assetRepo.Delete(id);
    }

    static async Update(asset: Asset): Promise<boolean> {
        return this.assetRepo.Update(asset);
    }

    static async GetAssetParentFacility(asset: Asset): Promise<Facility | null> {
        return await FacilityService.Get(asset.parentFacilityId);
    }

    static async GetAssetCurrentZones(asset: Asset): Promise<Zone[]> {
        let zones : Zone[] = [] ;
        for (const zoneId of asset.currentZonesIds){
            let zone = await ZoneService.Get(zoneId);
            if (zone != null){
                zones.push(zone);
            }
        }    
        return zones;
    }
}
